src/utils/interpolate.hpp

```diff
@@ -0,0 +1,134 @@
+#ifndef INTERPOLATE_HPP_
+#define INTERPOLATE_HPP_
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2022 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+//! \file interpolate.hpp
+//  \brief Interpolation using cubic Hermite polynomials
+//
+// Performs monotone interpolation using cubic Hermite polynomials.
+// Reference: "Monotone piecewise cubic interpolation," SIAM J. Numer. Anal.,
+//            Vol. 17, No. 2, April 1980. [https://doi.org/10.1137/0717021]
+//========================================================================================
+
+#include "../kokkos_abstraction.hpp"
+#include <ostream>
+
+template <class VectorContainer>
+class MonotoneInterpolator {
+ public:
+  using Real = typename VectorContainer::value_type;
+  MonotoneInterpolator(VectorContainer const &x, VectorContainer const &y);
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION
+  MonotoneInterpolator(MonotoneInterpolator const &rhs)
+      : x_min_(rhs.x_min_), x_max_(rhs.x_max_), x_vec_(rhs.x_vec_), f_vec_(rhs.f_vec_),
+        d_vec_(rhs.d_vec_) {}
+
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION auto operator()(Real x) const -> Real;
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION auto min() const -> Real { return x_min_; }
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION auto max() const -> Real { return x_max_; }
+  auto ConstructVectorContainer(size_t size) -> VectorContainer;
+
+ private:
+  Real x_min_{};
+  Real x_max_{}; // x_min and x_max are the min and max of x_vec_
+  VectorContainer x_vec_{};
+  VectorContainer f_vec_{};
+  VectorContainer d_vec_{};
+};
+
+template <>
+auto MonotoneInterpolator<std::vector<Real>>::ConstructVectorContainer(size_t size)
+    -> std::vector<Real> {
+  return std::vector<Real>(size);
+}
+
+template <>
+auto MonotoneInterpolator<PinnedArray1D<Real>>::ConstructVectorContainer(size_t size)
+    -> PinnedArray1D<Real> {
+  return PinnedArray1D<Real>("d", size);
+}
```

Comment:
I wonder if this can be generalized using @lroberts36 concepts so all view types could be specialized for at once.

Reply:
Currently `PinnedArray1D` is a typedef for `SharedHostPinnedSpace` in Kokkos 4, so this adds a requirement for Kokkos 4 if we have to explicitly specialize for common vector containers/views.

Reply:
I *think* this should just work, but Luke should comment
```C++
template<typename VectorContainer, class = ENABLEIF(implements<kokkos_view(VectorContainer)>::value)>
auto MonotoneInterpolator<VectorContainer>::ConstructVectorContainer(size_t size) {
  return VectorContainer("d", size);
}
```
you'll need to include `utils/concepts_lite.hpp`

Reply:
I think that should work, but it is probably more consistent to use `REQUIRES` instead  
```c++
template<typename VectorContainer, REQUIRES(implements<kokkos_view(VectorContainer)>::value)>
auto MonotoneInterpolator<VectorContainer>::ConstructVectorContainer(size_t size) {
  return VectorContainer("d", size);
}
```

Reply:
I can't get this to compile on my machine. I don't fully understand concepts, so I'm not sure why it's complaining:
```
/Users/benwibking/parthenon/src/utils/interpolate.hpp:54:45: error: nested name specifier 'MonotoneInterpolator<VectorContainer>::' for declaration does not refer into a class, class template or class template partial specialization
auto MonotoneInterpolator<VectorContainer>::ConstructVectorContainer(size_t size) {
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```